File: packages/core/stage-capsule-store.js
import crypto from "node:crypto";

import {
  ContractFault,
  domainContentRoot,
  validateClock,
  validateRoot,
} from "./canonical-contracts.js";
import { validateStageCapsule } from "./stage-capsule.js";

export const STAGE_CAPSULE_OUTPUT_MANIFEST_CONTRACT =
  "buildchain-v4-stage-capsule-output-manifest/v1";
export const STAGE_CAPSULE_RETENTION_STATE_CONTRACT =
  "buildchain-v4-stage-capsule-retention-state/v1";
export const STAGE_CAPSULE_TRANSPORT_CONTRACT =
  "buildchain-v4-stage-capsule-transport/v1";
export const STAGE_CAPSULE_STORE_RECEIPT_CONTRACT =
  "buildchain-v4-stage-capsule-store-receipt/v1";

const TOKEN = /^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/u;
const PROVIDERS = new Set([
  "local-filesystem",
  "github-artifacts",
  "s3-compatible",
]);
const TRANSPORT_MODES = new Set([
  "local-reference",
  "effect-disabled",
  "fixture-backed",
]);
const OPERATIONS = new Set(["put", "locate", "restore", "quarantine"]);
const OUTCOMES = new Set([
  "stored",
  "already-stored",
  "located",
  "restored",
  "quarantined",
]);

function fault(code, path, message) {
  throw new ContractFault(code, path, message);
}

function exactKeys(value, keys, path) {
  if (!value || typeof value !== "object" || Array.isArray(value))
    fault(
      "invalid-stage-capsule-store-shape",
      path,
      `${path} must be an object`,
    );
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (
    actual.length !== expected.length ||
    actual.some((key, index) => key !== expected[index])
  )
    fault(
      "invalid-stage-capsule-store-shape",
      path,
      `${path} keys are not canonical`,
    );
}

function token(value, path) {
  if (typeof value !== "string" || !TOKEN.test(value))
    fault(
      "invalid-stage-capsule-store-token",
      path,
      `${path} must be an ASCII token`,
    );
  return value;
}

export function stageCapsuleBlobRoot(bytes) {
  const value = Buffer.isBuffer(bytes) ? bytes : Buffer.from(bytes);
  return `sha256:${crypto.createHash("sha256").update(value).digest("hex")}`;
}

function manifestPayload(value) {
  return { schema: value.schema, entries: value.entries };
}

export function stageCapsuleOutputManifestRoot(value) {
  validateStageCapsuleOutputManifest(value, { verifyRoot: false });
  return domainContentRoot(
    "stage-capsule-output-manifest",
    manifestPayload(value),
  );
}

export function validateStageCapsuleOutputManifest(
  value,
  { verifyRoot = true } = {},
) {
  exactKeys(value, ["schema", "entries", "manifestRoot"], "$/manifest");
  if (value.schema !== STAGE_CAPSULE_OUTPUT_MANIFEST_CONTRACT)
    fault(
      "unsupported-stage-capsule-store-version",
      "$/manifest/schema",
      "unsupported output manifest schema",
    );
  if (!Array.isArray(value.entries) || value.entries.length === 0)
    fault(
      "invalid-stage-capsule-store-shape",
      "$/manifest/entries",
      "output manifest entries must be non-empty",
    );
  let prior = null;
  for (const [index, entry] of value.entries.entries()) {
    const entryPath = `$/manifest/entries/${index}`;
    exactKeys(entry, ["name", "root", "size"], entryPath);
    token(entry.name, `${entryPath}/name`);
    validateRoot(entry.root, `${entryPath}/root`);
    if (!Number.isSafeInteger(entry.size) || entry.size < 0)
      fault(
        "invalid-stage-capsule-content-size",
        `${entryPath}/size`,
        "content size must be a non-negative safe integer",
      );
    if (prior !== null && entry.name <= prior)
      fault(
        "unordered-stage-capsule-content",
        `${entryPath}/name`,
        "content names must be unique and byte-sorted",
      );
    prior = entry.name;
  }
  validateRoot(value.manifestRoot, "$/manifest/manifestRoot");
  if (
    verifyRoot &&
    value.manifestRoot !==
      domainContentRoot("stage-capsule-output-manifest", manifestPayload(value))
  )
    fault(
      "stage-capsule-manifest-root-mismatch",
      "$/manifest/manifestRoot",
      "manifestRoot does not bind the canonical output manifest",
    );
  return value;
}

export function stageCapsuleRetentionPromiseRoot(promise) {
  exactKeys(promise, ["class", "retainUntil"], "$/retentionPromise");
  token(promise.class, "$/retentionPromise/class");
  validateClock(promise.retainUntil, "$/retentionPromise/retainUntil");
  return domainContentRoot("stage-capsule-retention-promise", promise);
}

export function createStageCapsuleRetentionState({ capsule, evaluatedAt }) {
  validateStageCapsule(capsule);
  validateClock(evaluatedAt, "$/evaluatedAt");
  const value = {
    schema: STAGE_CAPSULE_RETENTION_STATE_CONTRACT,
    capsuleRoot: capsule.capsuleRoot,
    promiseRoot: stageCapsuleRetentionPromiseRoot(capsule.retentionPromise),
    evaluatedAt,
    status:
      evaluatedAt <= capsule.retentionPromise.retainUntil
        ? "retained"
        : "expired",
  };
  return {
    ...value,
    stateRoot: domainContentRoot("stage-capsule-retention-state", value),
  };
}

export function validateStageCapsuleRetentionState(value) {
  exactKeys(
    value,
    [
      "schema",
      "capsuleRoot",
      "promiseRoot",
      "evaluatedAt",
      "status",
      "stateRoot",
    ],
    "$/retentionState",
  );
  if (value.schema !== STAGE_CAPSULE_RETENTION_STATE_CONTRACT)
    fault(
      "unsupported-stage-capsule-store-version",
      "$/retentionState/schema",
      "unsupported retention-state schema",
    );
  validateRoot(value.capsuleRoot, "$/retentionState/capsuleRoot");
  validateRoot(value.promiseRoot, "$/retentionState/promiseRoot");
  validateClock(value.evaluatedAt, "$/retentionState/evaluatedAt");
  if (!new Set(["retained", "expired"]).has(value.status))
    fault(
      "invalid-stage-capsule-retention-state",
      "$/retentionState/status",
      "retention status must be retained or expired",
    );
  validateRoot(value.stateRoot, "$/retentionState/stateRoot");
  const { stateRoot, ...body } = value;
  if (stateRoot !== domainContentRoot("stage-capsule-retention-state", body))
    fault(
      "stage-capsule-retention-root-mismatch",
      "$/retentionState/stateRoot",
      "stateRoot does not bind the retention observation",
    );
  return value;
}

export function createStageCapsuleTransport({
  provider,
  mode,
  locatorRoot,
  observedAt,
}) {
  token(provider, "$/transport/provider");
  if (!PROVIDERS.has(provider) || !TRANSPORT_MODES.has(mode))
    fault(
      "unsupported-stage-capsule-transport",
      "$/transport",
      "unsupported transport provider or mode",
    );
  validateRoot(locatorRoot, "$/transport/locatorRoot");
  validateClock(observedAt, "$/transport/observedAt");
  const value = {
    schema: STAGE_CAPSULE_TRANSPORT_CONTRACT,
    provider,
    mode,
    locatorRoot,
    observedAt,
  };
  return {
    ...value,
    transportRoot: domainContentRoot("stage-capsule-transport", value),
  };
}

export function validateStageCapsuleTransport(value) {
  exactKeys(
    value,
    [
      "schema",
      "provider",
      "mode",
      "locatorRoot",
      "observedAt",
      "transportRoot",
    ],
    "$/transport",
  );
  const rebuilt = createStageCapsuleTransport(value);
  if (
    rebuilt.schema !== value.schema ||
    rebuilt.transportRoot !== value.transportRoot
  )
    fault(
      "stage-capsule-transport-root-mismatch",
      "$/transport/transportRoot",
      "transportRoot does not bind the transport observation",
    );
  return value;
}

export function createStageCapsuleStoreReceipt(value) {
  const body = {
    schema: STAGE_CAPSULE_STORE_RECEIPT_CONTRACT,
    operation: value.operation,
    recordedAt: value.recordedAt,
    capsuleRoot: value.capsuleRoot,
    manifestRoot: value.manifestRoot,
    retentionStateRoot: value.retentionStateRoot,
    availabilityRoot: value.availabilityRoot,
    transportRoot: value.transportRoot,
    qualificationRoot: value.qualificationRoot,
    outcome: value.outcome,
    faultCode: value.faultCode ?? null,
  };
  const receipt = {
    ...body,
    receiptRoot: domainContentRoot("stage-capsule-store-receipt", body),
  };
  return validateStageCapsuleStoreReceipt(receipt);
}

export function validateStageCapsuleStoreReceipt(value) {
  exactKeys(
    value,
    [
      "schema",
      "operation",
      "recordedAt",
      "capsuleRoot",
      "manifestRoot",
      "retentionStateRoot",
      "availabilityRoot",
      "transportRoot",
      "qualificationRoot",
      "outcome",
      "faultCode",
      "receiptRoot",
    ],
    "$/receipt",
  );
  if (value.schema !== STAGE_CAPSULE_STORE_RECEIPT_CONTRACT)
    fault(
      "unsupported-stage-capsule-store-version",
      "$/receipt/schema",
      "unsupported store receipt schema",
    );
  if (!OPERATIONS.has(value.operation) || !OUTCOMES.has(value.outcome))
    fault(
      "invalid-stage-capsule-store-receipt",
      "$/receipt",
      "unsupported store operation or outcome",
    );
  validateClock(value.recordedAt, "$/receipt/recordedAt");
  for (const name of [
    "capsuleRoot",
    "manifestRoot",
    "retentionStateRoot",
    "availabilityRoot",
    "transportRoot",
    "qualificationRoot",
    "receiptRoot",
  ])
    validateRoot(value[name], `$/receipt/${name}`);
  if (value.faultCode !== null)
    fault(
      "invalid-stage-capsule-store-receipt",
      "$/receipt/faultCode",
      "successful store receipts cannot carry a fault",
    );
  const { receiptRoot, ...body } = value;
  if (receiptRoot !== domainContentRoot("stage-capsule-store-receipt", body))
    fault(
      "stage-capsule-store-receipt-root-mismatch",
      "$/receipt/receiptRoot",
      "receiptRoot does not bind the store receipt",
    );
  return value;
}

function validateProviderRequest(operation, request) {
  const keys =
    operation === "put"
      ? ["capsuleRoot", "manifestRoot", "locatorRoot", "recordedAt"]
      : ["capsuleRoot", "locatorRoot", "recordedAt"];
  exactKeys(request, keys, "$/adapter/request");
  for (const name of keys.filter((name) => name.endsWith("Root")))
    validateRoot(request[name], `$/adapter/request/${name}`);
  validateClock(request.recordedAt, "$/adapter/request/recordedAt");
  return request;
}

export function createStageCapsuleProviderAdapter({
  provider,
  mode = "effect-disabled",
  fixture = null,
}) {
  if (!new Set(["github-artifacts", "s3-compatible"]).has(provider))
    fault(
      "unsupported-stage-capsule-transport",
      "$/provider",
      "provider adapter must target GitHub artifacts or S3-compatible storage",
    );
  if (!new Set(["effect-disabled", "fixture-backed"]).has(mode))
    fault(
      "unsupported-stage-capsule-transport",
      "$/mode",
      "provider adapter mode is unsupported",
    );
  if (mode === "fixture-backed" && (!fixture || typeof fixture !== "object"))
    fault(
      "invalid-stage-capsule-store-shape",
      "$/fixture",
      "fixture-backed adapters require explicit fixture operations",
    );
  const invoke = (operation, request) => {
    if (mode === "effect-disabled" || typeof fixture[operation] !== "function")
      fault(
        "stage-capsule-transport-disabled",
        "$/adapter",
        `${provider} ${operation} effects are disabled in this slice`,
      );
    const validated = validateProviderRequest(operation, request);
    return {
      authority: "fixture-only",
      provider,
      result: fixture[operation](validated),
    };
  };
  return Object.freeze({
    provider,
    mode,
    observe: (request) => {
      exactKeys(request, ["locatorRoot", "observedAt"], "$/adapter/request");
      return createStageCapsuleTransport({
        provider,
        mode,
        locatorRoot: request.locatorRoot,
        observedAt: request.observedAt,
      });
    },
    put: (request) => invoke("put", request),
    restore: (request) => invoke("restore", request),
  });
}

File: crates/buildchain-domain-contracts/src/stage_capsule_store.rs
use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::stage_capsule::{StageCapsule, StageCapsuleAvailability};
use crate::{
    ContractFault, ContractResult, ascii_token, content_root, validate_clock, validate_root,
};

pub const STAGE_CAPSULE_OUTPUT_MANIFEST_CONTRACT: &str =
    "buildchain-v4-stage-capsule-output-manifest/v1";
pub const STAGE_CAPSULE_RETENTION_STATE_CONTRACT: &str =
    "buildchain-v4-stage-capsule-retention-state/v1";
pub const STAGE_CAPSULE_TRANSPORT_CONTRACT: &str = "buildchain-v4-stage-capsule-transport/v1";
pub const STAGE_CAPSULE_STORE_RECEIPT_CONTRACT: &str =
    "buildchain-v4-stage-capsule-store-receipt/v1";
const STAGE_CAPSULE_STORE_FIXTURE_CONTRACT: &str = "buildchain-v4-stage-capsule-store-fixtures/v1";

fn fault(code: &str, path: &str, message: impl Into<String>) -> Box<ContractFault> {
    Box::new(ContractFault::validation(code, path, message))
}

pub fn stage_capsule_blob_root(bytes: &[u8]) -> String {
    format!("sha256:{:x}", Sha256::digest(bytes))
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StageCapsuleContentEntry {
    pub name: String,
    pub root: String,
    pub size: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StageCapsuleOutputManifest {
    pub schema: String,
    pub entries: Vec<StageCapsuleContentEntry>,
    pub manifest_root: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct ManifestPayload<'a> {
    schema: &'a str,
    entries: &'a [StageCapsuleContentEntry],
}

impl StageCapsuleOutputManifest {
    fn payload(&self) -> ManifestPayload<'_> {
        ManifestPayload {
            schema: &self.schema,
            entries: &self.entries,
        }
    }

    pub fn calculated_root(&self) -> ContractResult<String> {
        let value = serde_json::to_value(self.payload())
            .map_err(|error| fault("canonicalization-failed", "$/manifest", error.to_string()))?;
        content_root("stage-capsule-output-manifest", &value)
    }

    pub fn validate(&self) -> ContractResult<()> {
        if self.schema != STAGE_CAPSULE_OUTPUT_MANIFEST_CONTRACT || self.entries.is_empty() {
            return Err(fault(
                "unsupported-stage-capsule-store-version",
                "$/manifest",
                "unsupported or empty Stage Capsule output manifest",
            ));
        }
        let mut prior: Option<&str> = None;
        for (index, entry) in self.entries.iter().enumerate() {
            if !ascii_token(&entry.name) {
                return Err(fault(
                    "invalid-stage-capsule-store-token",
                    &format!("$/manifest/entries/{index}/name"),
                    "content name must be an ASCII token",
                ));
            }
            validate_root(&entry.root, &format!("$/manifest/entries/{index}/root"))?;
            if prior.is_some_and(|name| entry.name.as_str() <= name) {
                return Err(fault(
                    "unordered-stage-capsule-content",
                    &format!("$/manifest/entries/{index}/name"),
                    "content names must be unique and byte-sorted",
                ));
            }
            prior = Some(&entry.name);
        }
        validate_root(&self.manifest_root, "$/manifest/manifestRoot")?;
        if self.manifest_root != self.calculated_root()? {
            return Err(fault(
                "stage-capsule-manifest-root-mismatch",
                "$/manifest/manifestRoot",
                "manifestRoot does not bind the canonical output manifest",
            ));
        }
        Ok(())
    }
}

fn retention_promise_root(capsule: &StageCapsule) -> ContractResult<String> {
    let value = serde_json::to_value(&capsule.retention_promise).map_err(|error| {
        fault(
            "canonicalization-failed",
            "$/retentionPromise",
            error.to_string(),
        )
    })?;
    content_root("stage-capsule-retention-promise", &value)
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StageCapsuleRetentionState {
    pub schema: String,
    pub capsule_root: String,
    pub promise_root: String,
    pub evaluated_at: String,
    pub status: String,
    pub state_root: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct RetentionStatePayload<'a> {
    schema: &'a str,
    capsule_root: &'a str,
    promise_root: &'a str,
    evaluated_at: &'a str,
    status: &'a str,
}

impl StageCapsuleRetentionState {
    pub fn from_capsule(capsule: &StageCapsule, evaluated_at: &str) -> ContractResult<Self> {
        capsule.validate()?;
        validate_clock(evaluated_at, "$/retentionState/evaluatedAt")?;
        let mut state = Self {
            schema: STAGE_CAPSULE_RETENTION_STATE_CONTRACT.to_owned(),
            capsule_root: capsule.capsule_root.clone(),
            promise_root: retention_promise_root(capsule)?,
            evaluated_at: evaluated_at.to_owned(),
            status: if evaluated_at <= capsule.retention_promise.retain_until.as_str() {
                "retained".to_owned()
            } else {
                "expired".to_owned()
            },
            state_root: String::new(),
        };
        state.state_root = state.calculated_root()?;
        Ok(state)
    }

    fn payload(&self) -> RetentionStatePayload<'_> {
        RetentionStatePayload {
            schema: &self.schema,
            capsule_root: &self.capsule_root,
            promise_root: &self.promise_root,
            evaluated_at: &self.evaluated_at,
            status: &self.status,
        }
    }

    pub fn calculated_root(&self) -> ContractResult<String> {
        let value = serde_json::to_value(self.payload()).map_err(|error| {
            fault(
                "canonicalization-failed",
                "$/retentionState",
                error.to_string(),
            )
        })?;
        content_root("stage-capsule-retention-state", &value)
    }

    pub fn validate(&self) -> ContractResult<()> {
        if self.schema != STAGE_CAPSULE_RETENTION_STATE_CONTRACT
            || !["retained", "expired"].contains(&self.status.as_str())
        {
            return Err(fault(
                "invalid-stage-capsule-retention-state",
                "$/retentionState",
                "unsupported Stage Capsule retention state",
            ));
        }
        validate_root(&self.capsule_root, "$/retentionState/capsuleRoot")?;
        validate_root(&self.promise_root, "$/retentionState/promiseRoot")?;
        validate_clock(&self.evaluated_at, "$/retentionState/evaluatedAt")?;
        validate_root(&self.state_root, "$/retentionState/stateRoot")?;
        if self.state_root != self.calculated_root()? {
            return Err(fault(
                "stage-capsule-retention-root-mismatch",
                "$/retentionState/stateRoot",
                "stateRoot does not bind the retention observation",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StageCapsuleTransport {
    pub schema: String,
    pub provider: String,
    pub mode: String,
    pub locator_root: String,
    pub observed_at: String,
    pub transport_root: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct TransportPayload<'a> {
    schema: &'a str,
    provider: &'a str,
    mode: &'a str,
    locator_root: &'a str,
    observed_at: &'a str,
}

impl StageCapsuleTransport {
    fn payload(&self) -> TransportPayload<'_> {
        TransportPayload {
            schema: &self.schema,
            provider: &self.provider,
            mode: &self.mode,
            locator_root: &self.locator_root,
            observed_at: &self.observed_at,
        }
    }

    pub fn calculated_root(&self) -> ContractResult<String> {
        let value = serde_json::to_value(self.payload())
            .map_err(|error| fault("canonicalization-failed", "$/transport", error.to_string()))?;
        content_root("stage-capsule-transport", &value)
    }

    pub fn validate(&self) -> ContractResult<()> {
        if self.schema != STAGE_CAPSULE_TRANSPORT_CONTRACT
            || !["local-filesystem", "github-artifacts", "s3-compatible"]
                .contains(&self.provider.as_str())
            || !["local-reference", "effect-disabled", "fixture-backed"]
                .contains(&self.mode.as_str())
        {
            return Err(fault(
                "unsupported-stage-capsule-transport",
                "$/transport",
                "unsupported transport provider or mode",
            ));
        }
        validate_root(&self.locator_root, "$/transport/locatorRoot")?;
        validate_clock(&self.observed_at, "$/transport/observedAt")?;
        validate_root(&self.transport_root, "$/transport/transportRoot")?;
        if self.transport_root != self.calculated_root()? {
            return Err(fault(
                "stage-capsule-transport-root-mismatch",
                "$/transport/transportRoot",
                "transportRoot does not bind the transport observation",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StageCapsuleStoreReceipt {
    pub schema: String,
    pub operation: String,
    pub recorded_at: String,
    pub capsule_root: String,
    pub manifest_root: String,
    pub retention_state_root: String,
    pub availability_root: String,
    pub transport_root: String,
    pub qualification_root: String,
    pub outcome: String,
    pub fault_code: Option<String>,
    pub receipt_root: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct StoreReceiptPayload<'a> {
    schema: &'a str,
    operation: &'a str,
    recorded_at: &'a str,
    capsule_root: &'a str,
    manifest_root: &'a str,
    retention_state_root: &'a str,
    availability_root: &'a str,
    transport_root: &'a str,
    qualification_root: &'a str,
    outcome: &'a str,
    fault_code: &'a Option<String>,
}

impl StageCapsuleStoreReceipt {
    fn payload(&self) -> StoreReceiptPayload<'_> {
        StoreReceiptPayload {
            schema: &self.schema,
            operation: &self.operation,
            recorded_at: &self.recorded_at,
            capsule_root: &self.capsule_root,
            manifest_root: &self.manifest_root,
            retention_state_root: &self.retention_state_root,
            availability_root: &self.availability_root,
            transport_root: &self.transport_root,
            qualification_root: &self.qualification_root,
            outcome: &self.outcome,
            fault_code: &self.fault_code,
        }
    }

    pub fn calculated_root(&self) -> ContractResult<String> {
        let value = serde_json::to_value(self.payload())
            .map_err(|error| fault("canonicalization-failed", "$/receipt", error.to_string()))?;
        content_root("stage-capsule-store-receipt", &value)
    }

    pub fn validate(&self) -> ContractResult<()> {
        if self.schema != STAGE_CAPSULE_STORE_RECEIPT_CONTRACT
            || !["put", "locate", "restore", "quarantine"].contains(&self.operation.as_str())
            || ![
                "stored",
                "already-stored",
                "located",
                "restored",
                "quarantined",
            ]
            .contains(&self.outcome.as_str())
            || self.fault_code.is_some()
        {
            return Err(fault(
                "invalid-stage-capsule-store-receipt",
                "$/receipt",
                "unsupported successful store receipt",
            ));
        }
        validate_clock(&self.recorded_at, "$/receipt/recordedAt")?;
        for (root, path) in [
            (&self.capsule_root, "$/receipt/capsuleRoot"),
            (&self.manifest_root, "$/receipt/manifestRoot"),
            (&self.retention_state_root, "$/receipt/retentionStateRoot"),
            (&self.availability_root, "$/receipt/availabilityRoot"),
            (&self.transport_root, "$/receipt/transportRoot"),
            (&self.qualification_root, "$/receipt/qualificationRoot"),
            (&self.receipt_root, "$/receipt/receiptRoot"),
        ] {
            validate_root(root, path)?;
        }
        if self.receipt_root != self.calculated_root()? {
            return Err(fault(
                "stage-capsule-store-receipt-root-mismatch",
                "$/receipt/receiptRoot",
                "receiptRoot does not bind the store receipt",
            ));
        }
        Ok(())
    }
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct StoreFixtures {
    schema_version: u64,
    contract: String,
    valid_cases: Vec<StoreFixture>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct StoreFixture {
    id: String,
    capsule: StageCapsule,
    manifest: StageCapsuleOutputManifest,
    blobs: Vec<StoreFixtureBlob>,
    evaluated_at: String,
    retention_state: StageCapsuleRetentionState,
    transport: StageCapsuleTransport,
    availability: StageCapsuleAvailability,
    receipt: StageCapsuleStoreReceipt,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct StoreFixtureBlob {
    name: String,
    root: String,
    bytes_base64: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StoreFixtureProjection {
    id: String,
    manifest_root: String,
    retention_state_root: String,
    transport_root: String,
    availability_root: String,
    receipt_root: String,
    blob_roots: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StageCapsuleStoreFixtureProjection {
    pub valid_cases: Vec<StoreFixtureProjection>,
}

pub fn run_stage_capsule_store_fixture(
    bytes: &[u8],
) -> ContractResult<StageCapsuleStoreFixtureProjection> {
    let fixtures: StoreFixtures = serde_json::from_slice(bytes).map_err(|error| {
        fault(
            "invalid-stage-capsule-store-fixture",
            "$",
            error.to_string(),
        )
    })?;
    if fixtures.schema_version != 1 || fixtures.contract != STAGE_CAPSULE_STORE_FIXTURE_CONTRACT {
        return Err(fault(
            "unsupported-stage-capsule-store-version",
            "$/schemaVersion",
            "unsupported Stage Capsule store fixture contract",
        ));
    }
    let valid_cases = fixtures
        .valid_cases
        .into_iter()
        .map(|fixture| {
            fixture.capsule.validate()?;
            fixture.manifest.validate()?;
            if fixture.capsule.identity.output_manifest_root != fixture.manifest.manifest_root {
                return Err(fault(
                    "stage-capsule-root-mismatch",
                    "$/manifest/manifestRoot",
                    "capsule does not bind the fixture manifest",
                ));
            }
            if fixture.blobs.len() != fixture.manifest.entries.len() {
                return Err(fault(
                    "stage-capsule-partial",
                    "$/blobs",
                    "fixture blobs do not close the manifest",
                ));
            }
            let mut blob_roots = Vec::new();
            for (blob, entry) in fixture.blobs.iter().zip(&fixture.manifest.entries) {
                let decoded = BASE64.decode(&blob.bytes_base64).map_err(|_| {
                    fault(
                        "invalid-stage-capsule-store-fixture",
                        "$/blobs/bytesBase64",
                        "fixture blob is not canonical base64",
                    )
                })?;
                let root = stage_capsule_blob_root(&decoded);
                if blob.name != entry.name
                    || blob.root != entry.root
                    || root != entry.root
                    || decoded.len() as u64 != entry.size
                {
                    return Err(fault(
                        "stage-capsule-root-mismatch",
                        "$/blobs",
                        "fixture blob does not match the manifest",
                    ));
                }
                blob_roots.push(root);
            }
            let retention =
                StageCapsuleRetentionState::from_capsule(&fixture.capsule, &fixture.evaluated_at)?;
            retention.validate()?;
            fixture.retention_state.validate()?;
            if fixture.retention_state != retention {
                return Err(fault(
                    "stage-capsule-retention-root-mismatch",
                    "$/retentionState",
                    "fixture retention state is not the exact derived observation",
                ));
            }
            fixture.transport.validate()?;
            fixture.availability.validate()?;
            fixture.receipt.validate()?;
            let availability_root = fixture.availability.root()?;
            if fixture.receipt.capsule_root != fixture.capsule.capsule_root
                || fixture.receipt.manifest_root != fixture.manifest.manifest_root
                || fixture.receipt.retention_state_root != retention.state_root
                || fixture.receipt.transport_root != fixture.transport.transport_root
                || fixture.receipt.availability_root != availability_root
                || fixture.receipt.qualification_root != fixture.capsule.identity.qualification_root
            {
                return Err(fault(
                    "stage-capsule-root-mismatch",
                    "$/receipt",
                    "store receipt does not bind the fixture roots",
                ));
            }
            Ok(StoreFixtureProjection {
                id: fixture.id,
                manifest_root: fixture.manifest.manifest_root,
                retention_state_root: retention.state_root,
                transport_root: fixture.transport.transport_root,
                availability_root,
                receipt_root: fixture.receipt.receipt_root,
                blob_roots,
            })
        })
        .collect::<ContractResult<Vec<_>>>()?;
    Ok(StageCapsuleStoreFixtureProjection { valid_cases })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn raw_blob_roots_are_byte_sensitive() {
        assert_ne!(
            stage_capsule_blob_root(b"a"),
            stage_capsule_blob_root(b"a\n")
        );
    }
}
